### PR TITLE
Proactive vs. reactive content size batcher

### PR DESCRIPTION
**Pull request by swheaton**
## What changes are proposed in this pull request?

TBD

## How is this patch tested? If it is not, please explain why.

TBD

### CODE REVIEW

File: fiftyone/core/utils.py [hunk 1]
```diff
@@ -1539,7 +1534,102 @@ def _compute_batch_size(self):
...
+            callback function to invoke instead
+        total (None): the length of ``iterable``. Only applicable when
+            ``progress=True``. If not provided, it is computed via
+            ``len(iterable)``, if possible
+    """
+
+    def __init__(
+        self,
+        iterable,
+        size_calculation_fn,
```
Comment (CamronStaley) [inline fiftyone/core/utils.py:1584]: what's this size_calculation_fn? I don't see it in the comments or an example of it being passed in anywhere unless I am just missing it.

**Comment by CamronStaley** [comment]
workin on fixing that one test not sure why it's having issues yet

File: fiftyone/core/dataset.py [hunk 2]
```diff
@@ -3336,41 +3356,22 @@ def add_collection(
...
-            self._validate_samples(samples)
-
-        now = datetime.utcnow()
-        dicts = [
-            self._make_dict(sample, created_at=now, last_modified_at=now)
-            for sample in samples
-        ]
-
+    def _add_samples_batch(self, samples, batcher=None):
+        dicts = [sample[1] for sample in samples]
```
Comment (swheaton) [inline fiftyone/core/dataset.py:3360]: I was very confused until I realized samples is tuple(sample, transformed_dict).... this is where type hints would be 🤌

File: fiftyone/core/utils.py [hunk 3]
```diff
@@ -1138,19 +1138,24 @@ def __iter__(self):
                 )
                 self._render_progress = False
 
+        if self._transform_fn is not None:
+            self._iter = self._apply_transform(self._iter)
```
Comment (swheaton) [inline fiftyone/core/utils.py:1142]: could also just do `self._iter = map(self._transform_fn, self._iter)`

File: fiftyone/core/dataset.py [hunk 4]
```diff
@@ -3336,41 +3356,22 @@ def add_collection(
...
+        for sample, d in samples:
             doc = self._sample_dict_to_doc(d)
             sample._set_backing_doc(doc, dataset=self)
             if sample.media_type == fom.VIDEO:
                 sample.frames.save()
 
-        if batcher is not None and batcher.manual_backpressure:
-            # @todo can we infer content size from insert_many() above?
+        if batcher and batcher.manual_backpressure:
             batcher.apply_backpressure(dicts)
```
Comment (swheaton) [inline fiftyone/core/dataset.py:3375]: we don't need this any longer do we? esp. since you (or was it me i cant remember lol) moved it to Dynamic content size batcher only.

File: fiftyone/core/dataset.py [hunk 5]
```diff
@@ -3429,16 +3466,15 @@ def _upsert_samples_batch(
...
             sample._set_backing_doc(doc, dataset=self)
 
             if sample.media_type == fom.VIDEO:
                 sample.frames.save()
 
-        if batcher is not None and batcher.manual_backpressure:
-            # @todo can we infer content size from bulk_write() above?
-            batcher.apply_backpressure(dicts)
+        if batcher and batcher.manual_backpressure:
+            batcher.apply_backpressure([sample[1] for sample in samples])
```
Comment (swheaton) [inline fiftyone/core/dataset.py:3477]: same

File: fiftyone/core/utils.py [hunk 6]
```diff
@@ -1539,7 +1569,140 @@ def _compute_batch_size(self):
...
+                ):
+                    break
+                curr_batch_size += 1
+                batch_content_size += next_element_size
+                curr_batch.append(self._prev_element)
+            except StopIteration:
+                self._prev_element = None
+                break
+        self._last_batch_size = curr_batch_size
+        return curr_batch
```
Comment (swheaton) [inline fiftyone/core/utils.py:1684]: I think I'm the one that wrote this lol. But I'm wondering if there's a cleaner way, this is a little ugly 🤣

**Review by swheaton — COMMENTED** [review]
done with first pass, comments in line; gonna do a second pass in IDE

File: fiftyone/core/utils.py (hunk 6 above)
Comment (CamronStaley) [reply fiftyone/core/utils.py:1684]: yeah you wrote this lol I will try and refactor tmr

File: fiftyone/core/dataset.py (hunk 4 above)
Comment (CamronStaley) [reply fiftyone/core/dataset.py:3375]: is it not still needed for latency batcher? I will have to look on IDE in the morning but yeah can clean these up if it's no longer used outside of dynamic content size.

File: fiftyone/core/utils.py [hunk 7]
```diff
@@ -1539,7 +1569,140 @@ def _compute_batch_size(self):
...
+        return len(json_util.dumps(obj))
+    except Exception:
+        return len(str(obj))
+
+
+def get_default_batcher(
+    iterable,
+    progress=False,
+    total=None,
+    size_calc_fn=default_calculate_size,
```
Comment (swheaton) [inline fiftyone/core/utils.py:1703]: I think this can default None, since the default is handled in ContentSizeBatcher init, yes?

Also, its missing from docstring

File: fiftyone/core/utils.py [hunk 8]
```diff
@@ -1539,7 +1569,140 @@ def _compute_batch_size(self):
...
+
+    def __next__(self):
+        if self._render_progress and self._last_batch_size:
+            self._pb.update(count=self._last_batch_size)
+
+        if self._prev_element is None:
+            raise StopIteration()
+
+        batch_content_size = self._size_calculation_fn(self._prev_element)
+        curr_batch_size = 1
```
Comment (swheaton) [inline fiftyone/core/utils.py:1661]: unnecessary as we can just use len(curr_batch) where we need it

File: fiftyone/core/utils.py [hunk 9]
```diff
@@ -1433,6 +1453,7 @@ class ContentSizeDynamicBatcher(BaseDynamicBatcher):
         total (None): the length of ``iterable``. Only applicable when
             ``progress=True``. If not provided, it is computed via
             ``len(iterable)``, if possible
+        transform_fn (None): a transform function to apply to each item of the batch
     """
 
     manual_backpressure = True
```
Comment (swheaton) [inline fiftyone/core/utils.py:1459]: Yeah I think we can get rid of this variable entirely

File: fiftyone/core/utils.py [hunk 10]
```diff
@@ -1069,7 +1069,6 @@ def progress_dt(pb):
 
 
 class Batcher(abc.ABC):
```
Comment (swheaton) [inline fiftyone/core/utils.py:1071]: I don't like what I did with the class naming. How about this becomes `BaseBatcher` and `BaseBatcher` becomes something like `BaseChunkyBatcher`. Meaning the batch size is determined per chunk via `_compute_batch_size()`
Would have to explain this in docstring of course 🙃

**Comment by swheaton** [comment]
Awesome 👏🏼 . I'll test it out next. And re-review when you say go

File: fiftyone/core/dataset.py (hunk 4 above)
Comment (swheaton) [reply fiftyone/core/dataset.py:3375]: It is not. ContentSizeDynamicBatcher is the only one where manual_backpressure is True.
I had defined that variable on the root batcher since we can get 1 of a few different batchers, and I didn't want to do `hasattr(batcher, "manual_backpressure")`.
But that batcher isn't in get_default_batcher anymore so we can get rid of it

File: fiftyone/core/utils.py (hunk 6 above)
Comment (swheaton) [reply fiftyone/core/utils.py:1684]: ya i can help

File: fiftyone/core/dataset.py (hunk 2 above)
Comment (kaixi-wang) [reply fiftyone/core/dataset.py:3360]: Yeah can we add a comment at least?

File: fiftyone/core/utils.py (hunk 6 above)
Comment (swheaton) [reply fiftyone/core/utils.py]: failed to come up with something better than this so just renamed prev_elt to next_elt which makes more sense, and added some comments.

**Comment by swheaton** [comment]
LGTM! Pushed some minor coderabbit suggestions.
Also tested and it works! No readtimeout or duplicate key error.

🚢 🇮🇹 
(This is technically my PR so I cannot review it)

**Review by CamronStaley — APPROVED** [review]
